ckan/logic/validators.py

```diff
@@ -232,6 +232,43 @@ def resource_id_does_not_exist(
     if parent_id != package_id:
         errors[key].append(_('Resource id already exists.'))
 
+def group_id_does_not_exist(value: str, context: Context) -> Any:
+    """Ensures that the value is not used as a ID or name.
+    """
+
+    model = context['model']
+    session = context['session']
+
+    result = session.query(model.Group).get(value)
+    if result:
+        raise Invalid(_('Id already exists'))
+    return value
+
+def user_id_does_not_exist(value: str, context: Context) -> Any:
+    """Ensures that the value is not used as a ID or name.
+    """
+
+    model = context['model']
+    session = context['session']
+
+    result = session.query(model.User).get(value)
+    if result:
+        raise Invalid(_('Id already exists'))
+    return value
+
+
+def resource_view_id_does_not_exist(value: str, context: Context) -> Any:
+    """Ensures that the value is not used as a ID or name.
+    """
+
+    model = context['model']
+    session = context['session']
+
+    result = session.query(model.ResourceView).get(value)
+    if result:
+        raise Invalid(_('ResourceView id already exists'))
```

Comment:
Note that these don't completely protect us. If a user is sending multiple requests to create something with the same id it's possible for one to overwrite the other unless we pass a flag through to the save code to have postgres only create the record, on commit we'll reliably get an error if there was a conflict on one of the id fields. That's going to be too late for adding an error to the validation but at least we could prevent *bad things* from happening.

OTOH maybe all that's not worth the effort if the only users that can pass ids are sysadmins.

Reply:
Do you mean a user sending multiple requests to create new objects with the same new id? If so they will get a postgres PK error even if they bypass this right? What I mean is wouldn't our current code already do that?

Reply:
For example if two resources are created at the same time with the same id it's possible that
1. both pass validation at the same time (id doesn't exist)
2. one completes `resource_dict_save` to create the resource
3. the other begins `resource_dict_save` which turns into an update of the resource from step 2 because of https://github.com/ckan/ckan/blob/d77ceb7f0a4065d13781ed4673ba9a69ab7f9e97/ckan/lib/dictization/model_save.py#L24-L37

A fix is adding `create`/`update` parameters on all our `..._dict_save` functions (instead of this upsert-like behavior embedded in them) so that we will reliably a PK error in this case.

I could make a separate issue for this because this PR as it stands is a big improvement even without fixing our dict_save functions.

Reply:
Yeah let's do that in a separate PR. I'm sure we can include it in 2.11

Reply:
:+1: made an issue for it #8230